File: backend/src/main/java/com/university/application/service/EnrollmentService.java
package com.university.application.service;

import com.university.domain.model.enums.EnrollmentStatus;
import com.university.web.dto.enrollment.EnrollmentCreateRequestDTO;
import com.university.web.dto.enrollment.EnrollmentResponseDTO;
import com.university.web.dto.enrollment.EnrollmentUpdateRequestDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface EnrollmentService {

    //CRUD
    EnrollmentResponseDTO createEnrollment(EnrollmentCreateRequestDTO dto);
    EnrollmentResponseDTO getEnrollmentById(Long id);
    EnrollmentResponseDTO updateEnrollment(Long id, EnrollmentUpdateRequestDTO dto);
    void deleteEnrollment(Long id);

    // Búsquedas
    List<EnrollmentResponseDTO> getAllEnrollments();
    Page<EnrollmentResponseDTO> getAllEnrollmentsPageable(Pageable pageable);

    // Búsquedas estudiante
    List<EnrollmentResponseDTO> getEnrollmentsByStudent(Long studentId);
    List<EnrollmentResponseDTO> getEnrollmentsByStudentAndStatus(Long studentId, EnrollmentStatus status);
    List<EnrollmentResponseDTO> getEnrollmentsByStudentAndPeriod(Long studentId, Long periodId);

    // Búsquedas por oferta de curso
    List<EnrollmentResponseDTO> getEnrollmentsByCourseOffering(Long courseOfferingId);

    // Búsquedas por status
    List<EnrollmentResponseDTO> getEnrollmentsByStatus(EnrollmentStatus status);

    // Acciones especiales
    void withdrawEnrollment(Long id);
    void completeEnrollment(Long id);
    void approveEnrollment(Long id, Double finalGrade);
    void failEnrollment(Long id, Double finalGrade);
}

File: university-frontend/src/app/core/services/enrollment.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Enrollment, EnrollmentCreateRequest } from '../models/enrollment.model';

@Injectable({
  providedIn: 'root'
})
export class EnrollmentService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/enrollments`;

  createEnrollment(request: EnrollmentCreateRequest): Observable<Enrollment> {
    return this.http.post<Enrollment>(this.apiUrl, request);
  }

  getEnrollmentsByStudent(studentId: number): Observable<Enrollment[]> {
    return this.http.get<Enrollment[]>(`${this.apiUrl}/student/${studentId}`);
  }

  getEnrollmentsByStudentAndPeriod(studentId: number, periodId: number): Observable<Enrollment[]> {
    return this.http.get<Enrollment[]>(`${this.apiUrl}/student/${studentId}/period/${periodId}`);
  }

  withdrawEnrollment(id: number): Observable<Enrollment> {
    return this.http.patch<Enrollment>(`${this.apiUrl}/${id}/withdraw`, {});
  }

  updateEnrollment(id: number, request: any): Observable<Enrollment> {
    return this.http.put<Enrollment>(`${this.apiUrl}/${id}`, request);
  }

  deleteEnrollment(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }

  completeEnrollment(id: number): Observable<Enrollment> {
    return this.http.patch<Enrollment>(`${this.apiUrl}/${id}/complete`, {});
  }

  approveEnrollment(id: number, finalGrade: number): Observable<Enrollment> {
    return this.http.patch<Enrollment>(`${this.apiUrl}/${id}/approve?finalGrade=${finalGrade}`, {});
  }

  failEnrollment(id: number, finalGrade: number): Observable<Enrollment> {
    return this.http.patch<Enrollment>(`${this.apiUrl}/${id}/fail?finalGrade=${finalGrade}`, {});
  }

  getEnrollmentsByStatus(status: string): Observable<Enrollment[]> {
    return this.http.get<Enrollment[]>(`${this.apiUrl}/status/${status}`);
  }

  getEnrollmentsByCourseOffering(courseOfferingId: number): Observable<Enrollment[]> {
    return this.http.get<Enrollment[]>(`${this.apiUrl}/course-offering/${courseOfferingId}`);
  }

  getEnrollmentsByStudentAndStatus(studentId: number, status: string): Observable<Enrollment[]> {
    return this.http.get<Enrollment[]>(`${this.apiUrl}/student/${studentId}/status/${status}`);
  }

}   